QuickFIXn/DefaultMessageFactory.cs

```diff
@@ -119,41 +116,42 @@ public Group Create(string beginString, string msgType, int groupCounterTag)
             return dict;
         }
 
+        private static bool _dllsAreLoaded = false;
+        private static readonly object _dllLoadSync = new object();
+
         private static void LoadLocalDlls()
         {
-            const int @true = 1;
-
-            // Because we want to attempt load assemblies once only
-            var loadFlag = Interlocked.Exchange(ref _dllLoadFlag, @true);
-            if (loadFlag == @true)
-            {
+            if (_dllsAreLoaded)
                 return;
-            }
 
-            try
+            lock (_dllLoadSync)
             {
-                var assemblyLocation = Assembly.GetExecutingAssembly().Location;
-                if (String.IsNullOrWhiteSpace(assemblyLocation))
-                {
+                // check again in case the load happened while this thread was waiting for the lock
+                if (_dllsAreLoaded)
                     return;
-                }
 
-                var directory = Path.GetDirectoryName(assemblyLocation);
-                if (String.IsNullOrWhiteSpace(directory))
+                try
                 {
-                    return;
-                }
+                    var assemblyLocation = Assembly.GetExecutingAssembly().Location;
+                    if (String.IsNullOrWhiteSpace(assemblyLocation))
+                        return;
+
+                    var directory = Path.GetDirectoryName(assemblyLocation);
+                    if (String.IsNullOrWhiteSpace(directory))
+                        return;
+
+                    var dlls = Directory.GetFiles(directory, "QuickFix.*.dll");
+                    foreach (var path in dlls)
+                        Assembly.LoadFrom(path);
 
-                var dlls = Directory.GetFiles(directory, "QuickFix.*.dll");
-                foreach (var path in dlls)
+                    _dllsAreLoaded = true;
```

Comment:
I think for ultimate thread safety, either `_dllsAreLoaded` needs to be marked with `volatile` or the first `if` outside of the lock needs to be removed (easiest to reason about). I think otherwise there is nothing theoretically preventing this write being executed out of order and becoming observable (by another thread outside the lock) before the preceding code has finished.

Pretty theoretical and I may be wrong as to whether that can actually happen

Reply:
I think you're right that `volatile` is probably appropriate.

I was about to argue in favor of leaving the `if` in, but I realized that this function won't be called enough for that to cause any actual efficiency gain.  Not sure why I didn't realize that when I wrote it.